Retain result type after UNION as base type in T-SQL

hopebo opened the pull request:
### Description

When a base type such as `money` is UNIONed with another integer type, the result type will be `fixeddecimal` instead of `money`, which is inconsistent with SQL Server's behaviour. And could also cause some characters not displaying well due to inappropriate result type.
 
### Issues Resolved

The root cause of the issue is Postgres will consider the base type of domain type as the result type. In Babelfish, `money` is implemented as a domain type of `fixeddecimal`. Actually, `money` is a base type in T-SQL, thus Babelfish should retain `money` type after UNION.

After a final type has determined by the optimizer, check if it's a base type in T-SQL. If it is, try to use the domain type as result type if the domain type has higher precision than other types.

BABEL-5242
 
### Check List

- [x] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is under the terms of the PostgreSQL license, and grant any person obtaining a copy of the contribution permission to relicense all or a portion of my contribution to the PostgreSQL License solely to contribute all or a portion of my contribution to the PostgreSQL open source project.

For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/babelfish-for-postgresql/postgresql_modified_for_babelfish/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

forestkeeper commented on src/backend/parser/parse_coerce.c:
```diff
@@ -1507,32 +1508,80 @@ select_common_type(ParseState *pstate, List *exprs, const char *context,
+
+	if (sql_dialect == SQL_DIALECT_TSQL)
 	{
-		bool all_nullconst = true;
```
Is it possible to put all those logic into single hook and move to babelfish code only ?

hopebo replied:
I've optimized the codes and nest to the original loop. And it works together with the following codes outside of the loop. So it's not easy to make them in a single hook.

hopebo committed:
Retain result type after UNION as base type in T-SQL

Description
-----------
When a base type such as `money` is UNIONed with another integer type,
the result type will be `fixeddecimal` instead of `money`, which is
inconsistent with SQL Server's behaviour. And could also cause some
characters not displaying well.

Analysis
--------
The root cause of the issue is Postgres will consider the base type of
domain type as the result type. In Babelfish, `money` is implemented
as a domain type of `fixeddecimal`. Actually, `money` is a base type
in T-SQL, thus Babelfish should retain `money` type after UNION.

Fix
---
After a final type has determined by the optimizer, check if it's a
base type in T-SQL. If it is, try to use the domain type as result
type if the domain type has higher precision than other types.

Task: BABEL-5242
Signed-off-by: Bo Li <zboli@amazon.com>

Deepesh125 commented on src/backend/parser/parse_coerce.c:
```diff
@@ -1504,9 +1505,38 @@ select_common_type(ParseState *pstate, List *exprs, const char *context,
+				determine_datatype_precedence_hook &&
+				determine_datatype_precedence_hook(exprType(nexpr),
+												   exprType(pexpr)))
+				pexpr = nexpr;
```
why arent we updating ptype? because next iteration may need to make some decision

Deepesh125 replied:
second, why cant we fit this logic into select_common_type_hook?

hopebo replied:
Since in this branch, `ntype` is equal to `ptype`. So as `pcategory` and `pispreferred`. We change the `expr` for the information of domain type. This logic is combined with the following changes.

hopebo replied:
>second, why cant we fit this logic into select_common_type_hook?

It is because a general fix for this issue should happens after the main process in `select_common_type`, in which the common type has been decided. The `select_common_type_hook` does a early return. And if we want to do it in this hook, we need to copy most codes in `select_common_type` to that hook.
